File: python/mirage/utils/timezone.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import re
from collections.abc import Mapping
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone, tzinfo
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

TZ_VAR = "TZ"

# A POSIX TZ name: `<...>` quotes any run of letters, digits and signs
# (`<+0530>`), and a bare name is three or more letters (`EST`, `CEST`).
_NAME_RE = re.compile(r"<([+\-0-9A-Za-z]+)>|([A-Za-z]{3,})")
# A POSIX offset or rule time: `[+-]h[h[h]][:mm[:ss]]`.
_OFFSET_RE = re.compile(r"([+-]?)(\d{1,3})(?::(\d{1,2}))?(?::(\d{1,2}))?")
_RULE_RE = re.compile(r"M(\d{1,2})\.(\d)\.(\d)|J(\d{1,3})|(\d{1,3})")
_HOUR = 3600
_DAY = timedelta(days=1)
_WEEK = timedelta(days=7)
# The rule glibc applies when a DST name comes with no `,rule`: the US
# transitions, second Sunday of March and first Sunday of November.
_DEFAULT_RULES = "M3.2.0,M11.1.0"
# The window zone_abbreviations samples: tzdata's post-1970 rules, up to
# the 32-bit horizon every zone's rules are laid out to.
_ABBREV_SAMPLE_START = datetime(1970, 1, 1, tzinfo=timezone.utc)
_ABBREV_SAMPLE_END = datetime(2038, 1, 1, tzinfo=timezone.utc)


@dataclass(frozen=True, slots=True)
class TransitionRule:
    """One POSIX DST transition, ``Mm.w.d``, ``Jn`` or ``n``, with the
    local time of day it happens at.

    Args:
        kind (str): ``M`` for month/week/weekday, ``J`` for a Julian
            day that never counts February 29, ``D`` for a zero-based
            day of the year that does.
        month (int): the month for an ``M`` rule, 1 to 12.
        week (int): the week for an ``M`` rule, 1 to 5, where 5 is the
            last such weekday of the month.
        weekday (int): the weekday for an ``M`` rule, 0 for Sunday.
        day (int): the day for a ``J`` (1 to 365) or ``D`` (0 to 365)
            rule.
        seconds (int): the time of day, in seconds, which POSIX lets run
            past a day in either direction (``/-1``, ``/25``).
    """

    kind: str
    month: int = 0
    week: int = 0
    weekday: int = 0
    day: int = 0
    seconds: int = 2 * _HOUR

    def at(self, year: int) -> datetime:
        """The transition's wall-clock moment in ``year``.

        Args:
            year (int): the calendar year.
        """
        if self.kind == "M":
            first = datetime(year, self.month, 1)
            # Python counts Monday as 0; POSIX counts Sunday as 0.
            ahead = (self.weekday - (first.weekday() + 1)) % 7
            date = first + timedelta(days=ahead) + _WEEK * (self.week - 1)
            while date.month != self.month:
                date -= _WEEK
        elif self.kind == "J":
            date = datetime(year, 1, 1) + timedelta(days=self.day - 1)
            leap = year % 4 == 0 and (year % 100 != 0 or year % 400 == 0)
            if leap and self.day >= 60:
                date += _DAY
        else:
            date = datetime(year, 1, 1) + timedelta(days=self.day)
        return date + timedelta(seconds=self.seconds)


class PosixZone(tzinfo):
    """A zone read from a POSIX TZ string with a DST half
    (``CET-1CEST,M3.5.0,M10.5.0/3``), the way glibc reads one.

    The two offsets and the two rules decide everything: an instant is
    in DST when it lies between the start transition, given in standard
    wall time, and the end transition, given in DST wall time, with the
    window wrapping the year in the southern hemisphere. A wall clock
    that two instants share (the hour repeated at the end of DST)
    resolves to the first of them unless ``fold`` says otherwise, and
    one no instant shows (the hour skipped at its start) reads under
    the standard offset, which are Python's own conventions for a
    ``tzinfo``.

    Args:
        std (str): the standard-time abbreviation.
        std_offset (timedelta): the standard offset from UTC.
        dst (str): the daylight-time abbreviation.
        dst_offset (timedelta): the daylight offset from UTC.
        start (TransitionRule): when DST starts, in standard wall time.
        end (TransitionRule): when DST ends, in daylight wall time.
    """

    def __init__(self, std: str, std_offset: timedelta, dst: str,
                 dst_offset: timedelta, start: TransitionRule,
                 end: TransitionRule) -> None:
        self._std = std
        self._std_offset = std_offset
        self._dst = dst
        self._dst_offset = dst_offset
        self._start = start
        self._end = end

    def _in_dst(self, utc: datetime) -> bool:
        """Whether DST is in effect at a naive UTC moment.

        Args:
            utc (datetime): the moment, naive, on the UTC clock.
        """
        year = (utc + self._std_offset).year
        start = self._start.at(year) - self._std_offset
        end = self._end.at(year) - self._dst_offset
        if start < end:
            return start <= utc < end
        return not end <= utc < start

    def _offset_for(self, dt: datetime) -> timedelta:
        """The offset a wall-clock reading resolves to.

        Args:
            dt (datetime): the wall clock, whose ``fold`` picks the later
                of two instants sharing it.
        """
        wall = dt.replace(tzinfo=None)
        as_std = not self._in_dst(wall - self._std_offset)
        as_dst = self._in_dst(wall - self._dst_offset)
        if as_dst and (not as_std or dt.fold == 0):
            return self._dst_offset
        return self._std_offset

    def utcoffset(self, dt: datetime | None) -> timedelta:
        if dt is None:
            return self._std_offset
        return self._offset_for(dt)

    def dst(self, dt: datetime | None) -> timedelta:
        if dt is None or self._offset_for(dt) == self._std_offset:
            return timedelta(0)
        return self._dst_offset - self._std_offset

    def tzname(self, dt: datetime | None) -> str:
        if dt is not None and self._offset_for(dt) == self._dst_offset:
            return self._dst
        return self._std

    def fromutc(self, dt: datetime) -> datetime:
        utc = dt.replace(tzinfo=None)
        in_dst = self._in_dst(utc)
        wall = utc + (self._dst_offset if in_dst else self._std_offset)
        # The second reading of a repeated hour: a standard-time wall
        # clock that would also have resolved as daylight time.
        fold = int(not in_dst and self._in_dst(wall - self._dst_offset))
        return wall.replace(tzinfo=self, fold=fold)

    def __repr__(self) -> str:
        return (f"PosixZone({self._std!r}, {self._std_offset!r}, "
                f"{self._dst!r}, {self._dst_offset!r})")


def zone_from_env(env: Mapping[str, str] | None) -> tzinfo | None:
    """The zone a command environment's ``TZ`` names.

    None when ``TZ`` is unset (or there is no environment), which means
    the host's local zone, the way a naive ``datetime`` does. The value
    is read from the command's own environment, never from the process
    (``os.environ``, ``time.tzset``), so two workspaces running at once
    each see their own ``TZ`` and neither moves the host's clock.

    Args:
        env (Mapping[str, str] | None): the command environment.
    """
    if env is None:
        return None
    spec = env.get(TZ_VAR)
    if spec is None:
        return None
    return resolve_tz(spec)


def resolve_tz(spec: str) -> tzinfo:
    """The zone a ``TZ`` value names, read as glibc's ``tzset`` reads it.

    A leading colon is dropped. An empty value is UTC. A name tzdata
    knows (``Asia/Hong_Kong``, ``UTC``, ``EST5EDT``) is that zone.
    Anything else is a POSIX TZ string (``UTC0``, ``JST-9``,
    ``<+0530>-5:30``, ``CET-1CEST,M3.5.0,M10.5.0/3``), where a bare name
    with no offset is UTC under that name, which is how glibc renders
    ``TZ=Bogus/Zone`` (``+0000 Bogus``), and a name shorter than three
    letters is refused, leaving ``%Z`` empty.

    Args:
        spec (str): the value of ``TZ``.
    """
    name = spec[1:] if spec.startswith(":") else spec
    if not name:
        return timezone.utc
    try:
        return ZoneInfo(name)
    except (ZoneInfoNotFoundError, ValueError):
        return posix_zone(name)


def _read_name(spec: str, pos: int) -> tuple[str, int]:
    """Read a POSIX TZ name at ``pos``: the name and the position after
    it, or an empty name and ``pos`` when none starts there.

    Args:
        spec (str): the TZ string.
        pos (int): where to read.
    """
    match = _NAME_RE.match(spec, pos)
    if match is None:
        return "", pos
    return match.group(1) or match.group(2), match.end()


def _read_seconds(spec: str, pos: int) -> tuple[int | None, int]:
    """Read a POSIX ``[+-]hh[:mm[:ss]]`` at ``pos`` as signed seconds, or
    None when none starts there.

    Args:
        spec (str): the TZ string.
        pos (int): where to read.
    """
    match = _OFFSET_RE.match(spec, pos)
    if match is None:
        return None, pos
    sign = -1 if match.group(1) == "-" else 1
    hours = int(match.group(2))
    minutes = int(match.group(3) or 0)
    seconds = int(match.group(4) or 0)
    return sign * (hours * _HOUR + minutes * 60 + seconds), match.end()


def _read_rule(text: str) -> TransitionRule | None:
    """Read one ``,rule`` clause, or None when it is malformed.

    Args:
        text (str): the clause without its leading comma.
    """
    match = _RULE_RE.match(text)
    if match is None:
        return None
    seconds = 2 * _HOUR
    pos = match.end()
    if text[pos:pos + 1] == "/":
        read, pos = _read_seconds(text, pos + 1)
        if read is None:
            return None
        seconds = read
    if pos != len(text):
        return None
    if match.group(1) is not None:
        return TransitionRule("M",
                              month=int(match.group(1)),
                              week=int(match.group(2)),
                              weekday=int(match.group(3)),
                              seconds=seconds)
    if match.group(4) is not None:
        return TransitionRule("J", day=int(match.group(4)), seconds=seconds)
    return TransitionRule("D", day=int(match.group(5)), seconds=seconds)


def posix_zone(spec: str) -> tzinfo:
    """The zone a POSIX TZ string names.

    The grammar is ``std[offset[dst[offset][,start[/time],end[/time]]]]``.
    POSIX counts an offset west of Greenwich as positive, so ``EST5`` is
    five hours behind UTC; a missing offset is zero, and a missing
    daylight offset is one hour ahead of standard. A string that is not
    a TZ string at all (no name, or a rule that does not parse) is UTC
    with no abbreviation, which is what glibc falls back to.

    Args:
        spec (str): the TZ string, colon already dropped.
    """
    std, pos = _read_name(spec, 0)
    if not std:
        return timezone(timedelta(0), "")
    west, pos = _read_seconds(spec, pos)
    std_offset = timedelta(seconds=-(west or 0))
    dst, pos = _read_name(spec, pos)
    if not dst:
        return timezone(std_offset, std)
    west, pos = _read_seconds(spec, pos)
    dst_offset = (timedelta(
        seconds=-west) if west is not None else std_offset +
                  timedelta(hours=1))
    rules = spec[pos + 1:] if spec[pos:pos + 1] == "," else _DEFAULT_RULES
    clauses = rules.split(",")
    start = _read_rule(clauses[0])
    end = _read_rule(clauses[1]) if len(clauses) == 2 else None
    if start is None or end is None:
        return timezone(timedelta(0), "")
    return PosixZone(std, std_offset, dst, dst_offset, start, end)


def numeric_abbreviation(offset: int) -> str:
    """The abbreviation tzdata gives an offset it has no letters for.

    Since 2017 tzdata names such a zone by its offset, ``+08``, ``-03``
    or ``+0530``, with minutes only when they are not zero.

    Args:
        offset (int): the UTC offset in seconds, east positive.
    """
    sign = "-" if offset < 0 else "+"
    hours, rest = divmod(abs(offset), _HOUR)
    minutes = rest // 60
    return f"{sign}{hours:02d}" + (f"{minutes:02d}" if minutes else "")


def zone_abbreviations(name: str,
                       step: timedelta = _DAY) -> list[tuple[int, str]]:
    """The lettered abbreviations a tzdata zone has carried since 1970.

    Sampled once per ``step`` from 1970 to 2038, which catches every
    daylight period (the shortest lasts weeks) and every change of
    standard time. An abbreviation that is only the offset spelled out
    (``+08``) is left out, since ``numeric_abbreviation`` rebuilds it;
    when one offset has carried two names (Moscow's ``MSD`` summers
    before 2011 and its ``MSK`` at the same offset after them), the
    later one wins. This is the table the TypeScript twin ships for
    ``%Z``, because Intl offers no tzdata names.

    Args:
        name (str): a tzdata zone name.
        step (timedelta): the sampling interval.

    Returns:
        list[tuple[int, str]]: ``(offset_seconds, abbreviation)`` pairs,
        sorted by offset.
    """
    zone = ZoneInfo(name)
    latest: dict[int, str] = {}
    at = _ABBREV_SAMPLE_START
    while at < _ABBREV_SAMPLE_END:
        local = at.astimezone(zone)
        delta = local.utcoffset()
        abbrev = local.tzname()
        if delta is not None and abbrev is not None:
            latest[int(delta.total_seconds())] = abbrev
        at += step
    return sorted((offset, abbrev) for offset, abbrev in latest.items()
                  if abbrev != numeric_abbreviation(offset))

File: typescript/packages/core/src/utils/timezone.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

// The zone a command renders time in, read from its own environment's TZ
// the way glibc's tzset reads it, never from the process: two workspaces
// running at once each see their own TZ and neither moves the host's
// clock. Mirrors the Python mirage.utils.timezone.

import { TZ_ABBREVS } from './tz_abbrevs.ts'

export const TZ_VAR = 'TZ'

/** A wall-clock reading: the calendar fields a zone shows for an instant. */
export interface WallParts {
  year: number
  // 0 for January, as `Date` counts it.
  month: number
  day: number
  hour: number
  minute: number
  second: number
  ms: number
}

/** A wall-clock reading with the zone's offset and abbreviation at that instant. */
export interface ZoneParts extends WallParts {
  // 0 for Sunday.
  weekday: number
  // Seconds east of UTC.
  offsetSec: number
  abbrev: string
}

/**
 * A time zone: the wall clock it shows for an instant, and the instant its
 * wall clock reads as. `Date` has no zone of its own, so every rendering
 * (`strftime`) and every reading (`parseDateExpr`) goes through one of
 * these rather than through a `utc` flag.
 */
export interface Zone {
  parts(dt: Date): ZoneParts
  /**
   * The instant whose wall clock reads `p`. A wall clock two instants share
   * (the hour repeated when DST ends) resolves to the later one, as glibc's
   * mktime resolves it; one no instant shows (the hour skipped when DST
   * starts) resolves under the standard offset, so a caller that must
   * refuse it compares `parts` of the result against `p`.
   */
  fromWall(p: WallParts): Date
}

const HOUR = 3600
const DAY_MS = 86_400_000
const WEEK_MS = 7 * DAY_MS

// A POSIX TZ name: `<...>` quotes any run of letters, digits and signs
// (`<+0530>`), and a bare name is three or more letters (`EST`, `CEST`).
const NAME_RE = /<([+\-0-9A-Za-z]+)>|([A-Za-z]{3,})/y
// A POSIX offset or rule time: `[+-]h[h[h]][:mm[:ss]]`.
const OFFSET_RE = /([+-]?)(\d{1,3})(?::(\d{1,2}))?(?::(\d{1,2}))?/y
const RULE_RE = /M(\d{1,2})\.(\d)\.(\d)|J(\d{1,3})|(\d{1,3})/y
// The rule glibc applies when a DST name comes with no `,rule`: the US
// transitions, second Sunday of March and first Sunday of November.
const DEFAULT_RULES = 'M3.2.0,M11.1.0'

/**
 * A Date from wall-clock fields read as UTC. `Date.UTC` reads a year below
 * 100 as 1900 plus that year; the setters do not, so a year GNU and Python
 * accept as itself (`0042-01-01`) lands where it belongs. Overflowing
 * fields carry, so day 32 is the next month's first.
 */
export function utcFromWall(p: WallParts): Date {
  const d = new Date(0)
  d.setUTCFullYear(p.year, p.month, p.day)
  d.setUTCHours(p.hour, p.minute, p.second, p.ms)
  return d
}

function utcWall(dt: Date): WallParts & { weekday: number } {
  return {
    year: dt.getUTCFullYear(),
    month: dt.getUTCMonth(),
    day: dt.getUTCDate(),
    hour: dt.getUTCHours(),
    minute: dt.getUTCMinutes(),
    second: dt.getUTCSeconds(),
    ms: dt.getUTCMilliseconds(),
    weekday: dt.getUTCDay(),
  }
}

class UtcZone implements Zone {
  parts(dt: Date): ZoneParts {
    return { ...utcWall(dt), offsetSec: 0, abbrev: 'UTC' }
  }

  fromWall(p: WallParts): Date {
    return utcFromWall(p)
  }
}

// The host's zone, through the local `Date` getters. It carries no
// abbreviation: the runtime offers none short of Intl's default locale,
// and `%Z` has always rendered empty for a local time here.
class LocalZone implements Zone {
  parts(dt: Date): ZoneParts {
    return {
      year: dt.getFullYear(),
      month: dt.getMonth(),
      day: dt.getDate(),
      hour: dt.getHours(),
      minute: dt.getMinutes(),
      second: dt.getSeconds(),
      ms: dt.getMilliseconds(),
      weekday: dt.getDay(),
      offsetSec: -dt.getTimezoneOffset() * 60,
      abbrev: '',
    }
  }

  fromWall(p: WallParts): Date {
    const d = new Date(0)
    d.setFullYear(p.year, p.month, p.day)
    d.setHours(p.hour, p.minute, p.second, p.ms)
    return d
  }
}

/** One offset under one abbreviation: `UTC0`, `JST-9`, `<+0530>-5:30`. */
class FixedZone implements Zone {
  constructor(
    private readonly offsetSec: number,
    private readonly abbrev: string,
  ) {}

  parts(dt: Date): ZoneParts {
    const shown = utcWall(new Date(dt.getTime() + this.offsetSec * 1000))
    return { ...shown, offsetSec: this.offsetSec, abbrev: this.abbrev }
  }

  fromWall(p: WallParts): Date {
    return new Date(utcFromWall(p).getTime() - this.offsetSec * 1000)
  }
}

/**
 * The abbreviation tzdata gives an offset it has no letters for: `+08`,
 * `-03`, `+0530`, with minutes only when they are not zero.
 */
export function numericAbbreviation(offsetSec: number): string {
  const sign = offsetSec < 0 ? '-' : '+'
  const total = Math.abs(offsetSec)
  const hours = String(Math.floor(total / 3600)).padStart(2, '0')
  const minutes = Math.floor((total % 3600) / 60)
  return minutes === 0 ? `${sign}${hours}` : `${sign}${hours}${String(minutes).padStart(2, '0')}`
}

/**
 * The abbreviation tzdata gives `name` at `offsetSec`, which is what GNU
 * date prints for `%Z` and what Python's zoneinfo renders. Intl only
 * offers `GMT+8`, so the lettered names ship in TZ_ABBREVS (generated
 * from zoneinfo by scripts/gen_tz_abbrevs.py) and are looked up by the
 * offset Intl does report; an offset with no row is one tzdata spells
 * out (`+08`).
 */
export function tzAbbreviation(name: string, offsetSec: number): string {
  const row = TZ_ABBREVS[name]?.find(([offset]) => offset === offsetSec)
  return row === undefined ? numericAbbreviation(offsetSec) : row[1]
}

/**
 * A tzdata zone, read through Intl for its offsets and through
 * tzAbbreviation for its `%Z`, so `Asia/Hong_Kong` renders `HKT` here as
 * it does under GNU date and in the Python twin.
 */
class IntlZone implements Zone {
  private readonly format: Intl.DateTimeFormat

  constructor(private readonly name: string) {
    this.format = new Intl.DateTimeFormat('en-US', {
      timeZone: name,
      hourCycle: 'h23',
      year: 'numeric',
      month: 'numeric',
      day: 'numeric',
      hour: 'numeric',
      minute: 'numeric',
      second: 'numeric',
    })
  }

  parts(dt: Date): ZoneParts {
    const fields: Record<string, string> = {}
    for (const part of this.format.formatToParts(dt)) fields[part.type] = part.value
    const wall: WallParts = {
      year: Number(fields.year),
      month: Number(fields.month) - 1,
      day: Number(fields.day),
      hour: Number(fields.hour) % 24,
      minute: Number(fields.minute),
      second: Number(fields.second),
      ms: dt.getUTCMilliseconds(),
    }
    const shown = utcFromWall(wall)
    const offsetSec = Math.round((shown.getTime() - dt.getTime()) / 1000)
    return {
      ...wall,
      weekday: shown.getUTCDay(),
      offsetSec,
      abbrev: tzAbbreviation(this.name, offsetSec),
    }
  }

  fromWall(p: WallParts): Date {
    const wall = utcFromWall(p).getTime()
    // The offset at the wall clock read as UTC is within a day of the
    // right one; the offset at that first guess is the right one except
    // across a transition, where the two guesses are the two readings.
    const first = wall - this.parts(new Date(wall)).offsetSec * 1000
    const second = wall - this.parts(new Date(first)).offsetSec * 1000
    for (const t of [Math.max(first, second), Math.min(first, second)]) {
      if (utcFromWall(this.parts(new Date(t))).getTime() === wall) return new Date(t)
    }
    return new Date(second)
  }
}

/**
 * One POSIX DST transition, `Mm.w.d`, `Jn` or `n`, with the local time of
 * day it happens at: `M` is month/week/weekday (week 5 is the last such
 * weekday, weekday 0 is Sunday), `J` a Julian day that never counts
 * February 29, `D` a zero-based day of the year that does. The time may
 * run past a day in either direction (`/-1`, `/25`), as POSIX allows.
 */
export interface TransitionRule {
  kind: 'M' | 'J' | 'D'
  month: number
  week: number
  weekday: number
  day: number
  seconds: number
}

function isLeap(year: number): boolean {
  return year % 4 === 0 && (year % 100 !== 0 || year % 400 === 0)
}

/** The transition's wall-clock moment in `year`, as ms of that wall clock read as UTC. */
export function transitionAt(rule: TransitionRule, year: number): number {
  const jan1 = utcFromWall({ year, month: 0, day: 1, hour: 0, minute: 0, second: 0, ms: 0 })
  let date: number
  if (rule.kind === 'M') {
    const first = utcFromWall({
      year,
      month: rule.month - 1,
      day: 1,
      hour: 0,
      minute: 0,
      second: 0,
      ms: 0,
    })
    const ahead = (rule.weekday - first.getUTCDay() + 7) % 7
    date = first.getTime() + ahead * DAY_MS + (rule.week - 1) * WEEK_MS
    while (new Date(date).getUTCMonth() !== rule.month - 1) date -= WEEK_MS
  } else if (rule.kind === 'J') {
    date = jan1.getTime() + (rule.day - 1) * DAY_MS
    if (isLeap(year) && rule.day >= 60) date += DAY_MS
  } else {
    date = jan1.getTime() + rule.day * DAY_MS
  }
  return date + rule.seconds * 1000
}

/**
 * A zone read from a POSIX TZ string with a DST half
 * (`CET-1CEST,M3.5.0,M10.5.0/3`), the way glibc reads one. The two offsets
 * and the two rules decide everything: an instant is in DST when it lies
 * between the start transition, given in standard wall time, and the end
 * transition, given in DST wall time, with the window wrapping the year in
 * the southern hemisphere.
 */
class PosixZone implements Zone {
  constructor(
    private readonly std: string,
    private readonly stdOffsetSec: number,
    private readonly dst: string,
    private readonly dstOffsetSec: number,
    private readonly start: TransitionRule,
    private readonly end: TransitionRule,
  ) {}

  private inDst(utcMs: number): boolean {
    const year = new Date(utcMs + this.stdOffsetSec * 1000).getUTCFullYear()
    const start = transitionAt(this.start, year) - this.stdOffsetSec * 1000
    const end = transitionAt(this.end, year) - this.dstOffsetSec * 1000
    if (start < end) return start <= utcMs && utcMs < end
    return !(end <= utcMs && utcMs < start)
  }

  parts(dt: Date): ZoneParts {
    const inDst = this.inDst(dt.getTime())
    const offsetSec = inDst ? this.dstOffsetSec : this.stdOffsetSec
    const shown = utcWall(new Date(dt.getTime() + offsetSec * 1000))
    return { ...shown, offsetSec, abbrev: inDst ? this.dst : this.std }
  }

  fromWall(p: WallParts): Date {
    const wall = utcFromWall(p).getTime()
    const asStd = wall - this.stdOffsetSec * 1000
    const asDst = wall - this.dstOffsetSec * 1000
    // Standard time first: it is the later of two readings of a repeated
    // hour, which is the one glibc's mktime picks, and the reading a
    // skipped hour falls back to.
    if (!this.inDst(asStd)) return new Date(asStd)
    if (this.inDst(asDst)) return new Date(asDst)
    return new Date(asStd)
  }
}

export const UTC_ZONE: Zone = new UtcZone()
export const LOCAL_ZONE: Zone = new LocalZone()

/**
 * The zone a command environment's TZ names, or null when TZ is unset (or
 * there is no environment), which means the host's local zone.
 */
export function zoneFromEnv(env: Readonly<Record<string, string>> | null | undefined): Zone | null {
  const spec = env?.[TZ_VAR]
  if (spec === undefined) return null
  return resolveTz(spec)
}

/**
 * The zone a TZ value names, read as glibc's tzset reads it: a leading
 * colon is dropped; an empty value is UTC; a name tzdata knows
 * (`Asia/Hong_Kong`, `UTC`, `EST5EDT`) is that zone; anything else is a
 * POSIX TZ string (`UTC0`, `JST-9`, `<+0530>-5:30`,
 * `CET-1CEST,M3.5.0,M10.5.0/3`), where a bare name with no offset is UTC
 * under that name, which is how glibc renders `TZ=Bogus/Zone`
 * (`+0000 Bogus`), and a name shorter than three letters is refused,
 * leaving `%Z` empty.
 */
export function resolveTz(spec: string): Zone {
  const name = spec.startsWith(':') ? spec.slice(1) : spec
  if (name === '') return UTC_ZONE
  try {
    return new IntlZone(name)
  } catch (err) {
    if (!(err instanceof RangeError)) throw err
    return posixZone(name)
  }
}

function readName(spec: string, pos: number): [string, number] {
  NAME_RE.lastIndex = pos
  const m = NAME_RE.exec(spec)
  if (m === null) return ['', pos]
  return [m[1] ?? m[2] ?? '', NAME_RE.lastIndex]
}

// A POSIX `[+-]hh[:mm[:ss]]` at `pos` as signed seconds, or null when none starts there.
function readSeconds(spec: string, pos: number): [number | null, number] {
  OFFSET_RE.lastIndex = pos
  const m = OFFSET_RE.exec(spec)
  if (m === null) return [null, pos]
  const sign = m[1] === '-' ? -1 : 1
  const total = Number(m[2]) * HOUR + Number(m[3] ?? 0) * 60 + Number(m[4] ?? 0)
  return [sign * total, OFFSET_RE.lastIndex]
}

function readRule(text: string): TransitionRule | null {
  RULE_RE.lastIndex = 0
  const m = RULE_RE.exec(text)
  if (m === null) return null
  let seconds = 2 * HOUR
  let pos = RULE_RE.lastIndex
  if (text[pos] === '/') {
    const [read, after] = readSeconds(text, pos + 1)
    if (read === null) return null
    seconds = read
    pos = after
  }
  if (pos !== text.length) return null
  const base = { month: 0, week: 0, weekday: 0, day: 0, seconds }
  if (m[1] !== undefined) {
    return { ...base, kind: 'M', month: Number(m[1]), week: Number(m[2]), weekday: Number(m[3]) }
  }
  if (m[4] !== undefined) return { ...base, kind: 'J', day: Number(m[4]) }
  return { ...base, kind: 'D', day: Number(m[5]) }
}

/**
 * The zone a POSIX TZ string names:
 * `std[offset[dst[offset][,start[/time],end[/time]]]]`. POSIX counts an
 * offset west of Greenwich as positive, so `EST5` is five hours behind
 * UTC; a missing offset is zero, and a missing daylight offset is one hour
 * ahead of standard. A string that is not a TZ string at all (no name, or
 * a rule that does not parse) is UTC with no abbreviation, which is what
 * glibc falls back to.
 */
export function posixZone(spec: string): Zone {
  const [std, afterStd] = readName(spec, 0)
  if (std === '') return new FixedZone(0, '')
  const [west, afterOffset] = readSeconds(spec, afterStd)
  const stdOffsetSec = -(west ?? 0)
  const [dst, afterDst] = readName(spec, afterOffset)
  if (dst === '') return new FixedZone(stdOffsetSec, std)
  const [dstWest, afterDstOffset] = readSeconds(spec, afterDst)
  const dstOffsetSec = dstWest !== null ? -dstWest : stdOffsetSec + HOUR
  const rules = spec[afterDstOffset] === ',' ? spec.slice(afterDstOffset + 1) : DEFAULT_RULES
  const clauses = rules.split(',')
  const start = readRule(clauses[0] ?? '')
  const end = clauses.length === 2 ? readRule(clauses[1] ?? '') : null
  if (start === null || end === null) return new FixedZone(0, '')
  return new PosixZone(std, stdOffsetSec, dst, dstOffsetSec, start, end)
}
